### PR TITLE
rfcs: gpu: enable kernel reuse

### PR DESCRIPTION
**Pull request by rjoursler**
This is an RFC which targets improving oneDNN kernel reuse, especially for GPU kernels.

[Rendered Version](https://github.com/oneapi-src/oneDNN/blob/rjoursle/kernel_cache/rfcs/20230110-kernel-reuse/README.md)

**Edit:** After discussions with @densamoilov, updated to recommend Reuse Options 4: Create a Global Kernel Cache. The main reasoning is to avoid the restrictions previously required by Reuse Option 3. A new API option was added to Option 4 as well which proposes reusing the primitive cache capacity for the kernel cache capacity.

### CODE REVIEW

File: rfcs/20230110-kernel-reuse/README.md [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+curated set of kernels can be used to easily implement Reuse Option 2 which
+oneDNN users have expressed interest in. Option 1 is not being considered as it
+only handles OpenCL C kernels, which is insufficient for our needs.
+
+As a data point, a combination of Reuse Options 2 and 4 is exactly how the GEMM
+implementation works today. There are a finite number of curated GEMM
+implementations, but there are too many to store in a simple function registry
+though, so within oneMKL, these specializations are stored in a global kernel
+cache. In addition, the recommended design is part of what enabled matrix
+multiplication to support runtime dimensions in oneDNN.
```
Comment (petercad) [inline rfcs/20230110-kernel-reuse/README.md:271]: Quick comment/clarification: while GEMM kernels accept runtime dimensions, oneDNN GPU GEMM doesn't yet support runtime dimensions on the oneDNN API level (i.e. if users request runtime dimensions, it will go to a reference implementation).

File: rfcs/20230110-kernel-reuse/README.md [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+#### Pros:
+Allows user control.
+
+Only requires modifying OpenCL C kernels.
+
+#### Cons:
+Poor default behavior. The use of cl_cache requires intervention from our users
+to setup and manage the generated cache.
+
+No support for non-OpenCL C code generation.
```
Comment (echeresh) [inline rfcs/20230110-kernel-reuse/README.md:122]: What do you think about implementing a oneDNN version of `cl_cache`? It's a very attractive option as we basically need to implement a wrapper over `clBuildProgram` (or something like that). It's local, complexity is limited and we get the desired reuse for kernels out-of-the-box once the primitive is updated to use less shape-dependent macros.

Comment (rjoursler) [reply rfcs/20230110-kernel-reuse/README.md:122]: Any kernel caching behavior we implement should align with the primitive cache. As such, we should not implement a persistent cache like `cl_cache` here. Instead the behavior should match #1105. I think this would be a reasonable way to implement a global kernel cache, from Option 4, for OpenCL kernels, but I think there are better implementation options. 

First, cache lookup is a (somewhat) performance sensitive operation. As such, it is better to perform this lookup as soon as possible. In my opinion, [gpu_primitive_t::create_kernels](https://github.com/oneapi-src/oneDNN/blob/d49bfbe6c2c0725902b43b3d15c72cb3303be55d/src/gpu/gpu_primitive.hpp#L140) is the location where it makes the most sense for such a wrapper. An actual implementation could use the `kernel_ctx` and some implementation id (kernel name is not sufficient as multiple kernels may be compiled) as the lookup key to accomplish this with (nearly) the same complexity.

Second, I think completely divorcing kernel cache implementation from kernel implementations is a mistake. A minor issue is that the kernel cache should be opt-in as caching objects not designed for it will just waste resources. Second, the caching mechanism has two failure modes: false positives failures and false negatives. A wrapper around `clBluildProgram` provides very good resiliency against false positive failures and, IMO, is the main advantage of that design. False positives will be caught in our normal batched testing process though, while false negative are not caught by current testing processes. Because of this asymmetry, I think it is more important to optimize the design to catch false negatives. As what counts as a false negative is dependent on kernel implementation, this enforces (some) interdependency between the kernel cache and the implementation.

Comment (echeresh) [reply rfcs/20230110-kernel-reuse/README.md:122]: To clarify myself: when mentioning `cl_cache` I rather thought about having a single place where kernel caching is implemented, in a primitive-agnostic way.

I agree about performance conerns.

> A minor issue is that the kernel cache should be opt-in as caching objects not designed for it will just waste resources.

Do you have an example? We can start from the point where all kernels are cached, and gradually discuss and address exceptions.

> Second, the caching mechanism has two failure modes: false positives failures and false negatives. A wrapper around clBluildProgram provides very good resiliency against false positive failures and, IMO, is the main advantage of that design. False positives will be caught in our normal batched testing process though, while false negative are not caught by current testing processes. Because of this asymmetry, I think it is more important to optimize the design to catch false negatives. As what counts as a false negative is dependent on kernel implementation, this enforces (some) interdependency between the kernel cache and the implementation.

What do you mean by "false positive" and "false negative"?

> In my [prototype implementation](https://github.com/intel-innersource/libraries.performance.math.onednn/pull/4946), I created a new base class compile_params_t to address the above points. I created a new gpu_primitive_t::create_kernels interface which takes a compile_params_t structure (wrapped in a jit_function_generator_t) to enforce kernel cache opt-in. In addition, the implementation uses compile_params_t as a lookup key removing indirection to improve cache lookup performance. Finally, having a specific structure for compilation parameters will enable testing for false negatives on the compile_params_t structure.

Thanks. One thing that concerns me is that this implementation requires per-primitive updates. For the naïve/straightforward I would expect:

- To use 1) OpenCL file name(s) and 2) compilation option string as the key for the cache (cache is runtime, not persistent)
- Maybe rely on the primitive cache state if we need to take into account `ONEDNN_PRIMITIVE_CACHE_CAPACITY`
  - This also doesn't require per-primitive changes

Comment (rjoursler) [reply rfcs/20230110-kernel-reuse/README.md:122]: >>A minor issue is that the kernel cache should be opt-in as caching objects not designed for it will just waste resources.

>Do you have an example? We can start from the point where all kernels are cached, and gradually discuss and address exceptions.

This applies to essentially all OpenCL implementations except for gen9_eltwise. Currently, we are encoding enough problem specific parameters at compile time that a kernel cache provides, at best, the same resolution as the primitive cache. As one example, ref_eltwise encodes the global work size into compilation (see ref_eltwise.cpp:77). Because this is so ubiquitous, I think we should at least analyze an implementation for benefit before enabling caching.

 > What do you mean by "false positive" and "false negative"?
 
There are two main failure modes for a caching mechanism. False positive is the scenario that cache lookup succeeds, but does not find the correct implementation. False negative is the case where cache lookup fails even though the expected implementation is cached. False positive errors are generally caused by incorrect equality/hash behavior on the kernel cache's key. The false negative case are generally caused by passing unused/unnecessary compilation parameters or, as I have fixed in the past, passing uninitialized data. We should add tests to catch these failures modes, but any method of testing for unused/unnecessary parameters will, by necessity, be implementation dependent.
 
> Thanks. One thing that concerns me is that this implementation requires per-primitive updates. 

Thanks for the feedback. I completely agree on this concern, and I do address it in the prototype. The prototype just interprets the compile time parameters as a byte buffer and performs a hash/equality comparison on the byte buffer. This consolidates the kernel cache key's operation similar to your suggestion, but restricts compile time parameters as we need to initialize all data and avoid pointers members values. Currently, the only reason for the implementation specific functions is easy access to the structure size. I am open to changing this, it is simple to modify the kernel cache key's behavior. As this discussion is implementation specific, I think this discussion should continue in the the prototype PR, rather than in the RFC.

**Comment by echeresh** [comment]
> Because this is so ubiquitous, I think we should at least analyze an implementation for benefit before enabling caching.

I agree but I don't think this should necessarily impact the cache implementation choice.

> We should add tests to catch these failures modes, but any method of testing for unused/unnecessary parameters will, by necessity, be implementation dependent.

I agree on testing. However I see that we may have enough diagnostics without much assistance from the implementation side. Basically we can come up with a lot of test cases based on a single function to check whether the requested primitive/problem in the cache already:

```c++
clear_cache();
primitive_create(P1_params);
assert(in_cache(P2_params)); // This should hold if P1 and P2 params are similar enough <- this is implementation-specific
```

Analysis and refactoring of primitive implementations require a lot of work but if we want to test that externally we need to clearly define "classes" of similar primitives should must share the same kernel. Once this is defined, we don't care about the implementation side - it's up to tests to catch such false negatives.

File: rfcs/20230110-kernel-reuse/README.md (hunk 2 above)
Comment (echeresh) [reply rfcs/20230110-kernel-reuse/README.md:122]: > Because this is so ubiquitous, I think we should at least analyze an implementation for benefit before enabling caching.

I agree but I don't think this should necessarily impact the cache implementation choice.

> We should add tests to catch these failures modes, but any method of testing for unused/unnecessary parameters will, by necessity, be implementation dependent.

I agree on testing. However I see that we may have enough diagnostics without much assistance from the implementation side. Basically we can come up with a lot of test cases based on a single function to check whether the requested primitive/problem in the cache already:

```c++
clear_cache();
primitive_create(P1_params);
assert(in_cache(P2_params)); // This should hold if P1 and P2 params are similar enough <- this is implementation-specific
```

Analysis and refactoring of primitive implementations require a lot of work but if we want to test that externally we need to clearly define "classes" of similar primitives should must share the same kernel. Once this is defined, we don't care about the implementation side - it's up to tests to catch such false negatives.

Comment (rjoursler) [reply rfcs/20230110-kernel-reuse/README.md:122]: I agree with your assessment.  I am only mentioning these points to establish that any kernel cache implementation will require per kernel implementation work. We should strive to minimize that work. My takeaway is just that if a kernel cache implementation requires a few lines of change in each kernel implementation, this is not a significant problem. It will be significantly more work to bring implementations into a "testing/behavior compliant" state.

Comment (rjoursler) [reply rfcs/20230110-kernel-reuse/README.md:122]: Anyway disregarding implementation concerns, I do not have any issue with implementing a kernel cache. My main motivation against a kernel cache is that caching should only be used when necessary. In some sense, an implementation which requires a cache is a design flaw as there is always a usage scenario resulting in poor performance. As such, if we can provide kernel implementation which do not require a (dynamic) cache, that that the preferable solution.

File: rfcs/20230110-kernel-reuse/README.md [hunk 3]
```diff
@@ -0,0 +1,265 @@
+# Proposal for Primitive Compilation in oneDNN
+
+## 1. Introduction
+Slow primitive compilation consistently causes customer issues in oneDNN. To
+handle this, we introduced a [global primitive
+cache](https://github.com/oneapi-src/oneDNN/tree/rfcs/rfcs/20200207-global-primitive-cache).
+This solution is insufficient for the GPU software stack, especially for OpenCL
```
Comment (densamoilov) [inline rfcs/20230110-kernel-reuse/README.md:7]: Can you please elaborate on this a little bit? Basically, there are 3 use cases that I'm aware of:
1. OpenVINO use case. Framework creates primitives only once and then re-uses them. They reduces creation overhead by using persistent cache API. They don't rely on oneDNN primitive cache
2. PyTorch and TensorFlow use case. They create primitives for each call of an operation. The primitive cache is filled during the 1st iteration and then the primitives are re-used
3. In some cases frameworks use different problem sizes for each iteration which makes the primitive cache inefficient.

As far as I understand we reduced creation overhead for the first 2 uses cases. However, the 3rd one is still a problem. Given that, I think the problem statement should be focused on significant overhead for dynamic shapes specifically.

Comment (rjoursler) [reply rfcs/20230110-kernel-reuse/README.md:7]: Ok, I can expand on this, a couple of quick points.
>OpenVINO use case. Framework creates primitives only once and then re-uses them. They reduces creation overhead by using persistent cache API. They don't rely on oneDNN primitive cache.

To some extent, this is OpenVino making something work despite our libraries behavior. They also implement parallel kernel creation to workaround this too Our goal should be to have reasonable out of box behavior, not just provide tools to workaround poor behavior. As such, I am would not really categorize this as solving the problem. We are just providing a workaround.

> PyTorch and TensorFlow use case. They create primitives for each call of an operation. The primitive cache is filled during the 1st iteration and then the primitives are re-used

While this massively helps the issue, it is not enough. For example, we are by default compiling custom OpenCL kernels for each eltwise primitive. Some of these operations take around 0.04 ms to execute on the GPU. The compilation process takes about 500 ms. Because of this, our customers need to reuse that specific eltwise primtive 300 / 0.04 = 12,500 for compute time to even match the time invested on the CPU generating the primitive. Even if it hasn't been reported, it is *extremely* unlikely this is not a problem for some scenario. Preferably, we would not even call the OpenCL compiler for such short lived operations, but we need to resolve kernel reuse issues before we worry about alternative kernel distribution methods. While I don't think this is the most important issue, we should keep this first compilation issue in mind while designing a solution.

> In some cases frameworks use different problem sizes for each iteration which makes the primitive cache inefficient.

You are correct, this is the most pressing use case.

Comment (densamoilov) [reply rfcs/20230110-kernel-reuse/README.md:7]: > To some extent, this is OpenVino making something work despite our libraries behavior. They also implement parallel kernel creation to workaround this too Our goal should be to have reasonable out of box behavior, not just provide tools to workaround poor behavior. As such, I would not really categorize this as solving the problem. We are just providing a workaround.

I don't think I agree, the overhead is coming from using OpenCL programming model. There is nothing we can do to reduce compilation time if the kernels were not compiled before and stored in some cache so I don't think this is just a workaround. Another important thing to mention is that frameworks may define reasonable out-of-the-box behavior differently. For example, PyTorch and TF are fine with the primitive cache that is enabled out-of-the-box though OpenVINO doesn't need it but they need another feature which is persistent cache.

What I'm trying to say is that the problem seems to be solely about dynamic shapes to me and it should be investigated as such. To me this is a discussion about enabling runtime dimensions vs enabling kernel cache.

File: rfcs/20230110-kernel-reuse/README.md [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+* Reduces Testing Complexity
+
+While it is impossible to achieve this in general, for sufficiently large
+problems, there tend to only be a few important compiled configurations. Because
+of these advantages, I recommend we start curating a set of kernels which
+address most sufficiently large problems.
+
+To reuse, I recommend starting with Reuse Option 3, the fixed size function
+registry. The motivating reason is the kernels generated for Reuse Option 3 also
+satisfy the requirements for Distribution Option 1, enabling us to address
```
Comment (yiqianglee) [inline rfcs/20230110-kernel-reuse/README.md:268]: @rjoursler quick question, how option 3 address first compilation issue? 
You mention in option3 that generation happens on runtime, my understanding for the flow is: primitive creation -> get kernel generator if cached, create if not cached -> generate kernel, is it correct? 
If yes, for first compilation, we still see the overhead because kernel generator is not there, right? Anything I missed? Thanks!

Comment (rjoursler) [reply rfcs/20230110-kernel-reuse/README.md:268]: You are correct. In fact, there will be a first compilation issue with all the reuse options. The only way to completely remove first compilation for OpenCL kernels is to implement Distribution Option 1, where we move compilation to build time and distribute the binaries within the library.  This is currently infeasible as, at the very least, we would need to enable effective kernel reuse and simultaneously keep library a reasonable size.

Instead, the reuse options decrease the number of first compilations. The current implementation performs compilations on a per-primitive basis. For Options 3 and 4 the target is to compile on a per-kernel basis. This will allow us to reduce the number of first compilations to the point it is irrelevant for most use cases.

Comment (yiqianglee) [reply rfcs/20230110-kernel-reuse/README.md:268]: Thanks for explanation.

File: rfcs/20230110-kernel-reuse/README.md [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+no API changes. The only guarantee is that the registry is generated at runtime
+and will not use much memory until users creates primitives. It is expected the
+implementation will be a hashmap between a kernel configuration and a kernel
+generator. Registries will have a compile time defined maximum size to avoid
+excessive memory usage. As a registries size is dependent on implementation,
+separate registries will exist per implementation. As the requirements on this
+implementations are similar to Distribution Option 1,this can be used to enable
+build time compilation.
+
+### Reuse Option 4: Global Kernel Generator Cache
```
Comment (densamoilov) [inline rfcs/20230110-kernel-reuse/README.md:228]: I agree that this option looks like the best tradeoff. While the section has a sufficient high-level description of the option it lacks certain level of details about the design of the feature.
I think at least the following aspect should be covered:
* Definition of the key
* Definition of the value
* How/where load/store operations are performed
* Any data about primitive (`primitive_t`) creation cost without kernel creation. We need to make sure that the overhead is acceptable

Please see the following RFCs for reference:
* [Persistent cache](https://github.com/oneapi-src/oneDNN/tree/rfcs/rfcs/20210624-persistent-cache)
* [Optimization of primitive descriptor creation](https://github.com/oneapi-src/oneDNN/tree/rfcs/rfcs/20210129-optimize-pd-creation)
* [Library size reduction](https://github.com/oneapi-src/oneDNN/tree/rfcs/rfcs/20210510-binary-size-reduction)

A final note, when creating an RFC we usually try to provide details on design for all suggested options. I don't think it makes sense to do it now but I just wanted to mention it for future RFCs.
